src/Parallel/GlobalCache.hpp

```diff
@@ -634,16 +440,48 @@ template <typename GlobalCacheTag, typename Function, typename... Args>
 void GlobalCache<Metavariables>::mutate(const std::tuple<Args...>& args) {
   (void)Parallel::charmxx::RegisterGlobalCacheMutate<
       Metavariables, GlobalCacheTag, Function, Args...>::registrar;
-  if (mutable_global_cache_proxy_is_set()) {
-    // charm-aware version: Mutate the variable on all PEs on this node.
-    for (auto pe = CkNodeFirst(CkMyNode());
-         pe < CkNodeFirst(CkMyNode()) + CkNodeSize(CkMyNode()); ++pe) {
-      mutable_global_cache_proxy_[pe].template mutate<GlobalCacheTag, Function>(
-          args);
-    }
-  } else {
-    // version that bypasses proxies.  Just call the function.
-    mutable_global_cache_->template mutate<GlobalCacheTag, Function>(args);
+  using tag = MutableCacheTag<GlobalCache_detail::get_matching_mutable_tag<
+      GlobalCacheTag, Metavariables>>;
+
+  // Do the mutate.
+  std::apply(
+      [this](const auto&... local_args) {
+        // First mutex is for value of mutable tag
+        std::mutex& mutex = tuples::get<MutexTag<tag>>(mutexes_).first;
+        // Scoped for lock guard
+        {
+          const std::lock_guard<std::mutex> lock(mutex);
+          Function::apply(make_not_null(&(std::get<0>(
+                              tuples::get<tag>(mutable_global_cache_)))),
+                          local_args...);
```

Comment:
Should we dereference `unique_ptr`s here?  I guess not doing so is consistent with `db::mutate`, but you can't actually change what is being pointed to since that would violate the always-readable guarantee.

Reply:
I'd be in favor of dereferences so that people can't change things they shouldn't be able to